### PR TITLE
Use SWC to strip types for "--no-check" flag

### LINKED ISSUE #6864: Strip TypeScript types using SWC
as per title 

Waiting on:
https://github.com/dprint/dprint-plugin-typescript/pull/29
https://github.com/denoland/deno_lint/pull/239

### PR DESCRIPTION
**Pull request by bartlomieju**
`v1.2.1`

```
▶ ./third_party/prebuilt/mac/hyperfine -w 3 -i 'deno cache --reload --no-check std/examples/chat/server_test.ts'
Benchmark #1: deno cache --reload --no-check std/examples/chat/server_test.ts

  Time (mean ± σ):     979.5 ms ±  10.7 ms    [User: 1.640 s, System: 0.071 s]

  Range (min … max):   967.1 ms … 1003.7 ms

```

`this PR`
```
▶ ./third_party/prebuilt/mac/hyperfine -w 3 -i 'target/release/deno cache --reload --no-check std/examples/chat/server_test.ts'
Benchmark #1: target/release/deno cache --reload --no-check std/examples/chat/server_test.ts

  Time (mean ± σ):      90.3 ms ±   3.5 ms    [User: 69.6 ms, System: 16.7 ms]

  Range (min … max):    86.5 ms … 100.1 ms

```

Binary size increased from `55942928` to `58668472`

Closes #6864 

<s>Blocked by https://github.com/swc-project/swc/issues/903</s>

### CODE REVIEW

**Comment by bartlomieju** [comment]
@kitsonk I have a question for you - if we'll land this PR is anything stopping us from using TS compiler with `noEmit` option and doing the actual transpilation using SWC? If I'm not mistaken this move could shave off some time that is spent in TS compiler.

**Comment by kitsonk** [comment]
Couple thoughts... I know we are eager to do this, but I think the refactor of the compiler is important (#6894).  This should be even easier to implement when that is redone.

Second, we can't use it for the type checking without breaking a lot of code in the wild, especially code that doesn't use `export type` and `import type` plus `const enum` and potentially other code. (You want to run it with `"isolatedModules": true` so that the type checking is as close to just type stripping as possible). Honestly we wouldn't be able to do that until 2.0 IMO.

**Comment by bartlomieju** [comment]
> Couple thoughts... I know we are eager to do this, but I think the refactor of the compiler is important (#6894). This should be even easier to implement when that is redone.

I agree that compiler refactor is important, I don't suggest trying to hook SWC up to type checking process in this PR. Are you ok with landing change just for `--no-check`?

> Second, we can't use it for the type checking without breaking a lot of code in the wild, especially code that doesn't use export type and import type plus const enum and potentially other code. (You want to run it with "isolatedModules": true so that the type checking is as close to just type stripping as possible). Honestly we wouldn't be able to do that until 2.0 IMO.

@kitsonk thanks for clear explanation, that sounds good to me.

**Comment by kitsonk** [comment]
> I agree that compiler refactor is important, I don't suggest trying to hook SWC up to type checking process in this PR. Are you ok with landing change just for `--no-check`?

Yeah, I hadn't actually looked at the changes... straight forwards and should be easy to integrate. 👍 

Thinking more, we might want to help people make the transition.  I know we have avoided it, but if we surfaced `--isolated` as an option for `run` and `check`, then people would be able to easily run their code in type check and get back errors that are more meaningful to what they need to fix in order to support just type stripping.  It would make the introduction of the breaking change easier.

File: cli/swc_util.rs [hunk 1]
```diff
@@ -179,34 +190,77 @@ impl AstParser {
...
+          .to_writer(&mut buf)?;
+        let map = String::from_utf8(buf)?;
+
+        src.push_str("\n//# sourceMappingURL=data:application/json;base64,");
+        let encoded_map = base64::encode(map.as_bytes());
+        src.push_str(&encoded_map);
+      }
+      Ok(src)
+    })
+  }
```
Comment (ry) [inline cli/swc_util.rs:262]: Seems like it would be good to have a unit test for this function here.

File: cli/tsc.rs [hunk 2]
```diff
@@ -1725,7 +1712,7 @@ mod tests {
     let source_code = compiled_file.code;
     assert!(source_code
       .as_bytes()
-      .starts_with(b"console.log(\"Hello World\");"));
+      .starts_with(b"console.log('Hello World');"));
```
Comment (ry) [inline cli/tsc.rs:1715]: The output defaults to different quotes?

Comment (kitsonk) [reply cli/tsc.rs:1715]: Super yucky!  It really shouldn't be doing that.

Comment (bartlomieju) [reply cli/tsc.rs:1715]: In general the output from codegen is rather poorly formatted. I guess we can run it through prettier before caching to disk...

Comment (kdy1) [reply cli/tsc.rs:1715]: Well, swc_ecma_codegen sucks for a linter because it's implemented based on usage for the compiler.
I'll fix it asap.

Comment (bartlomieju) [reply cli/tsc.rs:1710]: Fixed after upgrade to latest `swc_ecma_codegen`

File: cli/tsc.rs [hunk 3]
```diff
@@ -742,8 +735,8 @@ impl TsCompiler {
 
   pub async fn transpile(
     &self,
-    global_state: GlobalState,
-    permissions: Permissions,
+    _global_state: GlobalState,
+    _permissions: Permissions,
```
Comment (ry) [inline cli/tsc.rs:739]: less complexity 👍

**Pull request merged**